src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -229,6 +222,97 @@ MultiLogicalPlanCreate(Query *originalQuery, Query *queryTree,
 }
 
 
+/*
+ * ShouldUseSubqueryPushDown determines whether it's desirable to use
+ * subquery pushdown to plan the query based on the original and
+ * rewritten query.
+ */
+static bool
+ShouldUseSubqueryPushDown(Query *originalQuery, Query *rewrittenQuery)
+{
+	/*
+	 * We check the existence of subqueries in FROM clause on the modified query
+	 * given that if postgres already flattened the subqueries, MultiPlanTree()
+	 * can plan corresponding distributed plan.
+	 */
+	if (SubqueryEntryList(rewrittenQuery) != NIL)
+	{
+		return true;
+	}
+
+	/*
+	 * We also check the existence of subqueries in WHERE clause. Note that
+	 * this check needs to be done on the original query given that
+	 * standard_planner() may replace the sublinks with anti/semi joins and
+	 * MultiPlanTree() cannot plan such queries.
+	 */
+	if (SublinkList(originalQuery) != NIL)
+	{
+		return true;
+	}
+
+	/*
+	 * We process function RTEs as subqueries, since the join order planner
+	 * does not know how to handle them.
+	 */
+	if (FindNodeCheck((Node *) originalQuery, IsFunctionRTE))
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * IsFunctionRTE determines whether the given node is a function RTE.
+ */
+static bool
+IsFunctionRTE(Node *node)
+{
+	if (IsA(node, RangeTblEntry))
+	{
+		RangeTblEntry *rangeTblEntry = (RangeTblEntry *) node;
+
+		if (rangeTblEntry->rtekind == RTE_FUNCTION)
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * FindNodeCheck finds a node for which the check function returns true.
+ */
+static bool
+FindNodeCheck(Node *node, bool (*check)(Node *))
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (check(node))
+	{
+		return true;
+	}
+
+	if (IsA(node, RangeTblEntry))
```

Comment:
Some notes/thoughts on `range_table_walker ()`: In postgres source, it is used very few times. I think because if you have `query_tree_walker ()` and always start with a `Query` node, you don't need `range_table_walker ()`. It mostly seems useful if you could directly call the function with `rte` parameters.

That said, I don't see any downside of keeping it here, just wanted to shard my understanding on the function.

Reply:
Actually, the right thing to do here is `return false`, because `query_tree_walker` already calls `range_table_walker` and it's being called redundantly here.

Reply:
you're right that calling this function with an RTE would cause it to not descend, but you can call it `range_table_walker(rte, FindNodeCheck, check, QTW_EXAMINE_RTES)` directly in that case